# Regression with a Tabular California Housing Dataset  

This project uses a tabular dataset derived from California housing data to predict housing prices. The dataset provides an excellent opportunity to practice regression techniques, feature engineering, and model evaluation.

[Download the dataset here](https://www.kaggle.com/competitions/playground-series-s3e1/data).


In [2]:
import pandas as pd
import numpy as np

from xgboost import XGBRegressor
from sklearn.metrics import mean_squared_error
from sklearn.model_selection import train_test_split


In [4]:
train = pd.read_csv("train.csv")
test = pd.read_csv("test.csv")

In [6]:
X = train.drop(columns=['id', 'MedHouseVal'])
y = train['MedHouseVal']

In [8]:
# Test Data
X_test = test.drop(columns=['id'])

In [10]:
best_params = {
    'colsample_bytree': 0.7367518666865607,
    'gamma': 0.04589953290672094,
    'learning_rate': 0.03824709648056803,
    'max_depth': 8,
    'min_child_weight': 4,
    'n_estimators': 303,
    'reg_alpha': 0.0017161101831750236,
    'reg_lambda': 1.6923737499050842,
    'subsample': 0.9227651908203118
}

In [14]:
model = XGBRegressor(random_state=42, **best_params)
model.fit(X, y)

XGBRegressor(base_score=None, booster=None, callbacks=None,
             colsample_bylevel=None, colsample_bynode=None,
             colsample_bytree=0.7367518666865607, device=None,
             early_stopping_rounds=None, enable_categorical=False,
             eval_metric=None, feature_types=None, gamma=0.04589953290672094,
             grow_policy=None, importance_type=None,
             interaction_constraints=None, learning_rate=0.03824709648056803,
             max_bin=None, max_cat_threshold=None, max_cat_to_onehot=None,
             max_delta_step=None, max_depth=8, max_leaves=None,
             min_child_weight=4, missing=nan, monotone_constraints=None,
             multi_strategy=None, n_estimators=303, n_jobs=None,
             num_parallel_tree=None, random_state=42, ...)

In [16]:
y_train_pred = model.predict(X)

# Calculate RMSE
mse = mean_squared_error(y, y_train_pred)
rmse = np.sqrt(mse)
print(f"RMSE: {rmse:.4f}")

RMSE: 0.4319


In [18]:
y_pred = model.predict(X_test)

In [20]:
submission = pd.DataFrame({
    'id': test['id'],
    'MedHouseVal': y_pred
})

submission.to_csv('submission.csv', index=False)

### Conclusion

In this regression project, the model was evaluated using the **Root Mean Squared Error (RMSE)** metric, which measures the difference between predicted and actual values. After training the model with the California Housing dataset using **XGBoost**, the RMSE on the training set was found to be **0.4319**.

The Kaggle score of **0.56748** suggests that the model performs well, but there might still be room for improvement, particularly in fine-tuning hyperparameters or incorporating additional features. Overall, the model has demonstrated a good balance between prediction accuracy and model complexity.
